README.md
# brettle:accounts-add-service

[![Build Status](https://travis-ci.org/brettle/meteor-accounts-add-service.svg?branch=master)](https://travis-ci.org/brettle/meteor-accounts-add-service)

Allow users to add login services to their accounts.

This package is part of the `brettle:accounts-*` suite of packages. See
[`brettle:accounts-deluxe`](https://atmospherejs.com/brettle/accounts-deluxe)
for an overview of the suite and a live demo.

## Features

- When a logged in user attempts to login using credentials that don't belong to
  an existing user, this package adds those credentials to the logged in user,
  so they are available for future logins. Any other top-level or profile
  properties that would have belonged to a new user with those credentials are
  also non-destructively merged into the existing user.

- When a logged in user tries to login using credentials that belong to an
  existing user, the user gets logged in as the existing user (unless that is
  otherwise prevented, for example by brettle:accounts-logout-to-switch).

- Works with any login service (accounts-password, acccounts-google, etc.)

- Works with `accounts-ui` and other similar packages.

- Does not permanently monkey patch Meteor core.

## Installation
```sh
meteor add brettle:accounts-add-service
```

## Usage

Nothing to do. It should just work once installed.

When a service is added to an existing user, it is reported as a
`Meteor.Error(Accounts.LoginCancelledError, 'New login not needed. Service will
be added to logged in user.')` because a new account was not created.  The
`accounts-ui` package and the `useraccounts` suite will ignore that error when
adding an external login service, but _not_ when adding the password service.
When adding the password service, they will display the "New login not
needed..." message as a red error message to the user. To suppress those error
messages, `meteor add brettle:accounts-patch-ui`.

## Notes

This package adds a `hasLoggedIn` property to a user's account when he logs in,
and does a one-time migration to add the same property to existing users that
have the `resume` service created by the Meteor accounts system when a user logs
in. This is to ensure that a user counts as having logged in even if his
`resume` service is later removed (e.g. to force him to re-authenticate for
security reasons).

The migration runs from a `Meteor.startup` handler. It uses an internal
`Mongo.Collection` to keep track of whether it already started. This ensures
that it runs at most once and doesn't run on more than one server. If for some
reason you need to disable the migration, use:

```js
AccountsAddService.databaseMigrationEnabled = false;
```


CHANGELOG.md
# Change Log
This file documents all notable changes to this project. 
This project adheres to [Semantic Versioning](http://semver.org/).

## [Unreleased]

## [1.0.0] 2015-12-13

### Changed

Moved the fix for brettle/meteor-accounts-deluxe#1 (Red text error message when adding password account) into optional `brettle:accounts-patch-ui` package which patches the `accounts-ui` package and the `useraccounts` suite to workaround the problem.

## [0.3.5] 2015-11-13

### Fixed

Bug #1: Existing account without resume service deleted if logged in user logs
into it. For details, see the [Notes section of README.md](README.md#notes)

## [0.3.4] 2015-11-09

### Fixed

Bug #3: Missing email expiration tokens

## [0.3.3] 2015-09-26

### Fixed

Support Meteor 1.2.

## [0.3.2] 2015-09-26

### Fixed

Support Meteor 1.2.

## [0.3.1] 2015-09-07

### Fixed

Bug brettle/meteor-accounts-deluxe#1: Red text error message when using
`accounts-ui` to add password account.

## 0.3.0 2015-08-17

### Changed

Worked around meteor/meteor#4331 and moved all workarounds to their own
packages.

### Added

Support Meteor 1.0.4 or later.

[Unreleased]: https://github.com/brettle/meteor-accounts-add-service/compare/v1.0.0...HEAD
[1.0.0]: https://github.com/brettle/meteor-accounts-add-service/compare/v0.3.5...v1.0.0
[0.3.5]: https://github.com/brettle/meteor-accounts-add-service/compare/v0.3.4...v0.3.5
[0.3.4]: https://github.com/brettle/meteor-accounts-add-service/compare/v0.3.3...v0.3.4
[0.3.3]: https://github.com/brettle/meteor-accounts-add-service/compare/v0.3.2...v0.3.3
[0.3.2]: https://github.com/brettle/meteor-accounts-add-service/compare/v0.3.1...v0.3.2
[0.3.1]: https://github.com/brettle/meteor-accounts-add-service/compare/v0.3.0...v0.3.1


accounts-add-service-tests.js
Tinytest.add('AccountsAddService - logged out user logging in succeeds', function (test) {
  var connection = DDP.connect(Meteor.absoluteUrl());

  Meteor.users.remove({ 'services.test1.name': "testname" });
  var testId = connection.call('login', { test1: "testname" }).id;
  test.isNotUndefined(testId);
  test.isNotNull(testId);
});

Tinytest.add('AccountsAddService - logged in user logging in as new user adds service and merges profile', function (test) {
  var connection = DDP.connect(Meteor.absoluteUrl());

  Meteor.users.remove({ 'services.test1.name': "testname"});
  var testId = connection.call('login', {
    test1: "testname",
    docDefaults: {
      username: 'testname'
    }
  }).id;
  test.isNotUndefined(testId);
  test.isNotNull(testId);
  var user = Meteor.users.findOne(testId);
  test.isUndefined(user.services.test2);

  Meteor.users.remove({ 'services.test2.name': "test2name"});
  test.throws(function () {
    connection.call('login', {
      test2: "test2name",
      docDefaults: {
        emails: [ { address: 'test2@example.com' } ],
        username: 'test2name',
        profile: { name: 'test2name' }
      }
    }).id;
  }, 'Service will be added');

  user = Meteor.users.findOne(testId);
  test.isNotUndefined(user.services.test2, 'user.services.test2 defined');
  test.isNotNull(user.services.test2, 'user.services.test2 not null');
  test.equal(user.profile.name, 'test2name', 'merge profile');
  test.equal(user.emails[0].address, 'test2@example.com', 'merge top-level');
  test.equal(user.username, 'testname', 'merge top-level non-destructively');

  connection.call('logout');
  var test2Id = connection.call('login', { test2: "test2name" }).id;
  test.equal(test2Id, testId, 'ids match');
});

Tinytest.add('AccountsAddService - logged in user logging in as existing user switches to that user', function (test) {
  var connection = DDP.connect(Meteor.absoluteUrl());

  Meteor.users.remove({ 'services.test2.name': "test2name"});
  var test2Id = connection.call('login', { test2: "test2name" }).id;
  test.isNotUndefined(test2Id);
  test.isNotNull(test2Id);

  connection.call('logout');
  Meteor.users.remove({ 'services.test1.name': "testname"});
  var testId = connection.call('login', { test1: "testname" }).id;
  test.isNotUndefined(testId);
  test.isNotNull(testId);
  var user = Meteor.users.findOne(testId);
  test.equal(user.services.test1.name, 'testname');

  var newTest2Id = connection.call('login', { test2: "test2name" }).id;
  test.notEqual(newTest2Id, testId, 'not test id');
  test.equal(newTest2Id, test2Id, 'same as original user');

  user = Meteor.users.findOne(newTest2Id);
  test.isUndefined(user.services.test1, 'no user.services.test1');
});


workaround-issue-4331-tests.js
if (Meteor.isClient) {
  Tinytest.addAsync('Workaround4331 - Multiple onLogin callbacks', function (test, done) {
    var onLogin1Called;
    var onLogin1Stopper = Accounts.onLogin(function () {
      onLogin1Stopper.stop();
      onLogin1Called = true;
    });
    var onLogin2Called;
    var onLogin2Stopper = Accounts.onLogin(function () {
      onLogin2Stopper.stop();
      onLogin2Called = true;
      test.isTrue(onLogin1Called, 'onLogin1 not called');
      test.isTrue(onLogin2Called, 'onLogin2 not called');
      done();
    });
    Accounts.callLoginMethod({
      methodArguments: [{
        test1: "test1-" + Random.id()
      }]
    });
  });
  Tinytest.addAsync('Workaround4331 - Multiple onLoginFailure callbacks', function (test, done) {
    var onLoginFailure1Called;
    var onLoginFailure1Stopper = Accounts.onLoginFailure(function () {
      onLoginFailure1Stopper.stop();
      onLoginFailure1Called = true;
    });
    var onLoginFailure2Called;
    var onLoginFailure2Stopper = Accounts.onLoginFailure(function () {
      onLoginFailure2Stopper.stop();
      onLoginFailure2Called = true;
      test.isTrue(onLoginFailure1Called, 'onLoginFailure1 not called');
      test.isTrue(onLoginFailure2Called, 'onLoginFailure2 not called');
      done();
    });
    Accounts.callLoginMethod({
      methodArguments: [{
        bogusLoginService: "test1-" + Random.id()
      }]
    });
  });
}


workaround-issue-4970-tests.js
Tinytest.addAsync('Workaround4970 - Workaround meteor bug #4970',
  function(test, done) {
    // Make sure we get logged in again if we reconnect after adding a service
    Accounts.callLoginMethod({
      methodArguments: [{
        test1: "test1-" + Random.id()
      }],
      userCallback: onUser1LoggedIn
    });

    function onUser1LoggedIn(err) {
      test.isUndefined(err, 'Unexpected error logging in as first user');
      test.isNotNull(Accounts.connection.onReconnect, 'onReconnect is null after first user logged in');
      Accounts.callLoginMethod({
        methodArguments: [{
          test2: "test2-" + Random.id()
        }],
        userCallback: onUser2Added
      });
    }

    var onLoginStopper;
    var timeoutHandle;
    function onUser2Added(err) {
      test.instanceOf(err, Meteor.Error, "Did not receive Meteor.error");
      test.equal(err.reason, AccountsAddService._mergeUserErrorReason);
      test.isNotNull(Accounts.connection.onReconnect, 'onReconnect is null');
      onLoginStopper = Accounts.onLogin(loggedInAfterReconnect);
      timeoutHandle = Meteor.setTimeout(failedToLoginAfterReconnect, 1000);
      Meteor.disconnect();
      Meteor.reconnect();
    }

    function loggedInAfterReconnect() {
      Meteor.clearTimeout(timeoutHandle);
      onLoginStopper.stop();
      done();
    }

    function failedToLoginAfterReconnect() {
      onLoginStopper.stop();
      test.fail('Failed to login after reconnecting.');
      done();
    }
  }
);


accounts-add-service.js
AccountsAddService = {};

function isMergeable(user) {
  // A user should be merged if they have never logged in. If they have
  // never logged in, they won't have a "resume" service.
  return !(user.services && user.services.resume);
}

var mergeUserErrorReason = 'New login not needed. Service will be added to logged in user.';

AccountsAddService._init = function () {
  AccountsMultiple.register(addServiceCallbackSet);
}

var addServiceCallbackSet = {
  validateSwitch: function(attemptingUser, attempt) {
    if (isMergeable(attempt.user)) {
      throw new Meteor.Error(Accounts.LoginCancelledError.numericError, mergeUserErrorReason);
    }
    return true;
  },
  onSwitchFailure: function(attemptingUser, failedAttempt) {
    if (!failedAttempt.error ||
      !failedAttempt.error.error || !failedAttempt.error.reason ||
      failedAttempt.error.error !== Accounts.LoginCancelledError.numericError ||
      failedAttempt.error.reason !== mergeUserErrorReason) {
      return;
    }
    var serviceName = failedAttempt.type;
    var serviceData = failedAttempt.user.services[serviceName];

    // Repin any pinned oauth credentials to the logged in user
    if (serviceName !== "password" && serviceName !== "resume")
      repinCredentials(serviceData, failedAttempt.user._id, attemptingUser._id);

    Meteor.users.remove(failedAttempt.user._id);

    // Copy the serviceData into Meteor.user.services[serviceName]
    var setAttrs = {};
    _.each(serviceData, function(value, key) {
      setAttrs["services." + serviceName + "." + key] = value;
    });

    // Non-destructively merge profile properties
    var attemptingProfile = attemptingUser.profile || {};
    var attemptProfile = failedAttempt.user.profile;
    attemptProfile = _.omit(attemptProfile, _.keys(attemptingProfile));
    _.each(attemptProfile, function(value, key) {
      setAttrs["profile." + key] = value;
    });

    // Non-destructively merge top-level properties
    var attemptingTop = attemptingUser || {};
    var attemptTop = failedAttempt.user;
    attemptTop = _.omit(attemptTop, _.keys(attemptingTop));
    delete attemptTop.profile; // handled above
    delete attemptTop.services; // handled above
    _.each(attemptTop, function(value, key) {
      setAttrs[key] = value;
    });

    Meteor.users.update(attemptingUser._id, {
      $set: setAttrs
    });
  }
};

var OAuthEncryption = Package["oauth-encryption"] && Package["oauth-encryption"].OAuthEncryption;

function repinCredentials(serviceData, oldUserId, newUserId) {
  _.each(_.keys(serviceData), function(key) {
    var value = serviceData[key];
    if (OAuthEncryption && OAuthEncryption.isSealed(value))
      value = OAuthEncryption.seal(OAuthEncryption.open(value, oldUserId), newUserId);
    serviceData[key] = value;
  });
}

AccountsAddService._init();


workaround-issue-4331.js
// Workaround Meteor issue #4331 for releases < 1.2:
// https://github.com/meteor/meteor/issues/4331
//
// Do nothing for versions of Accounts which don't require this workaround
if (Accounts._onLoginHook) {
  return;
}

// Maintain our own set of hooks
var onLoginHook = new Hook({
  debugPrintExceptions: 'Workaround4331 onLogin callback'
});

var onLoginFailureHook = new Hook({
  debugPrintExceptions: 'Workaround4331 onLoginFailure callback'
});

// Monkey patch Accounts.onLogin and Accounts.onLoginFailure to register the
// callbacks with our hooks instead of the ones maintained by Accounts.
var accountsOnLogin = Accounts.onLogin;
Accounts.onLogin = function(func) {
  return onLoginHook.register(func);
};
var accountsOnLoginFailure = Accounts.onLoginFailure;
Accounts.onLoginFailure = function(func) {
  return onLoginFailureHook.register(func);
};

// Monkey patch Accounts.callLoginMethod to wrap the userCallback option with a
// function which calls our hooks before calling the userCallback.
var accountsCallLoginMethod = Accounts.callLoginMethod;
Accounts.callLoginMethod = function (options) {
  options = _.clone(options);
  var origUserCallback = options.userCallback || function () {};
  options.userCallback = function (error) {
    if (!error) {
      onLoginHook.each(function (callback) {
        callback();
        return true;
      });
    } else {
      onLoginFailureHook.each(function (callback) {
        callback();
        return true;
      });
    }
    return origUserCallback.apply(this, arguments);
  };
  return accountsCallLoginMethod.call(this, options);
};


accounts-add-service-server.js
/* globals AccountsAddService, AccountsMultiple
*/
"use strict";
var mergeUserErrorReason = AccountsAddService._mergeUserErrorReason;

AccountsAddService.databaseMigrationEnabled = true;

AccountsAddService._migrationControl = 
  new Mongo.Collection("AccountsAddService._migrationControl");

// returns false if the migration isn't run, or the number of users updated
// if it is.
AccountsAddService._migrateDatabase = function () {
  if (! AccountsAddService.databaseMigrationEnabled) {
    return false;
  }
  var addHasLoggedIn = 
    AccountsAddService._migrationControl.findOne('addHasLoggedIn');
  if (addHasLoggedIn && addHasLoggedIn.startedAt) {
    return false;
  }
  if (! addHasLoggedIn) {
    try {
      AccountsAddService._migrationControl.insert({ _id: 'addHasLoggedIn' });
    } catch (err) {
      // Ignore duplicate key error thrown if already id already exists due to
      // concurrent insertion attempts
      if (! (err.name === 'MongoError' && err.message.indexOf('E11000'))) {
        throw err;
      }
    }
  }
  var numAffected = AccountsAddService._migrationControl.update({
    _id: 'addHasLoggedIn',
    startedAt: { $exists: false }
   }, {
    $set: {
      startedAt: new Date(),
    }
  });
  // Only one server will return numAffected !== 0. When numAffect === 0,
  // The migration was already started (and possibly finished)
  if (numAffected === 0) {
    return false;
  }
  
  var selector = { 
    hasLoggedIn: { $exists: false }, 
    'services.resume': { $exists: true }
  };
  var numUsersUpdated = Meteor.users.update(selector, {
    $set: { hasLoggedIn: true }
  });
  if (numUsersUpdated > 0) {
    console.log('brettle:accounts-add-service set hasLoggedIn = true for ' +
      numUsersUpdated + ' existing user(s).');  
  }
  AccountsAddService._migrationControl.update({
    _id: 'addHasLoggedIn',
   }, {
    $set: {
      finishedAt: new Date(),
    }
  });
  return numUsersUpdated;
};

Meteor.startup(AccountsAddService._migrateDatabase);

// The first time a user logs in, we set his hasLoggedIn property so that
// we don't accidentally merge his account if his "resume" service is removed.
Accounts.onLogin(function (attemptInfo) {
  if (attemptInfo.user && ! attemptInfo.user.hasLoggedIn) {
    Meteor.users.update(attemptInfo.user._id, {
      $set: { hasLoggedIn: true }
    });
  }
});

function isMergeable(user) {
  // A user should be merged if they have never logged in. If they have
  // never logged in, they won't have a "resume" service and they won't have
  // the `hasLoggedIn` property.
  return !(user.services && user.services.resume) && !(user.hasLoggedIn);
}

AccountsAddService._init = function () {
  AccountsMultiple.register(addServiceCallbackSet);
};

var addServiceCallbackSet = {
  validateSwitch: function(attemptingUser, attempt) {
    if (isMergeable(attempt.user)) {
      throw new Meteor.Error(Accounts.LoginCancelledError.numericError,
        mergeUserErrorReason);
    }
    return true;
  },
  onSwitchFailure: function(attemptingUser, failedAttempt) {
    if (!failedAttempt.error ||
      !failedAttempt.error.error || !failedAttempt.error.reason ||
      failedAttempt.error.error !== Accounts.LoginCancelledError.numericError ||
      failedAttempt.error.reason !== mergeUserErrorReason) {
      return;
    }
    var serviceName = failedAttempt.type;
    var serviceData = failedAttempt.user.services[serviceName];

    // Repin any pinned oauth credentials to the logged in user
    if (serviceName !== "password" && serviceName !== "resume") {
      repinCredentials(serviceData, failedAttempt.user._id, attemptingUser._id);
    }

    Meteor.users.remove(failedAttempt.user._id);

    // Copy the serviceData into Meteor.user.services[serviceName]
    var setAttrs = {};
    _.each(serviceData, function(value, key) {
      setAttrs["services." + serviceName + "." + key] = value;
    });

    // Non-destructively merge profile properties
    var attemptingProfile = attemptingUser.profile || {};
    var attemptProfile = failedAttempt.user.profile;
    attemptProfile = _.omit(attemptProfile, _.keys(attemptingProfile));
    _.each(attemptProfile, function(value, key) {
      setAttrs["profile." + key] = value;
    });

    // Non-destructively merge emails
    if (failedAttempt.user.emails) {
      var attemptingEmails = attemptingUser.emails || [];
      var attemptEmails = failedAttempt.user.emails;
      var mergedEmails = attemptingEmails.concat(attemptEmails);
      mergedEmails = _.uniq(mergedEmails, false, function (email) { 
        return email.address;
      });
      setAttrs.emails = mergedEmails;
    }

    // Non-destructively merge email verification tokens
    if (failedAttempt.user.services && 
        failedAttempt.user.services.email && 
        failedAttempt.user.services.email.verificationTokens) {
      var attemptingTokens = (
        attemptingUser.services && 
        attemptingUser.services.email && 
        attemptingUser.services.email.verificationTokens
      ) || [];
      var attemptTokens = failedAttempt.user.services.email.verificationTokens;
      var mergedTokens = attemptingTokens.concat(attemptTokens);
      mergedTokens = _.uniq(mergedTokens, false, function (tokenRecord) { 
        return tokenRecord.token;
      });
      setAttrs["services.email.verificationTokens"] = mergedTokens;
    }

    // Non-destructively merge top-level properties
    var attemptingTop = attemptingUser || {};
    var attemptTop = failedAttempt.user;
    attemptTop = _.omit(attemptTop, _.keys(attemptingTop));
    delete attemptTop.profile; // handled above
    delete attemptTop.emails; // handled above
    delete attemptTop.services; // handled above
    _.each(attemptTop, function(value, key) {
      setAttrs[key] = value;
    });

    Meteor.users.update(attemptingUser._id, {
      $set: setAttrs
    });
  }
};

var OAuthEncryption = Package["oauth-encryption"] &&
  Package["oauth-encryption"].OAuthEncryption;

function repinCredentials(serviceData, oldUserId, newUserId) {
  _.each(_.keys(serviceData), function(key) {
    var value = serviceData[key];
    if (OAuthEncryption && OAuthEncryption.isSealed(value)) {
      value =
        OAuthEncryption.seal(OAuthEncryption.open(value, oldUserId), newUserId);
    }
    serviceData[key] = value;
  });
}

AccountsAddService._init();


package.js
"use strict";

Package.describe({
  name: 'brettle:accounts-add-service',
  version: '1.0.0',
  summary: 'Allow users to add login services to their accounts',
  git: 'https://github.com/brettle/meteor-accounts-add-service.git',
  documentation: 'README.md'
});

Package.onUse(function(api) {
  api.versionsFrom('1.0.4');
  api.use('underscore');
  api.use('mongo');
  api.use('brettle:workaround-issue-4331@0.0.1');
  api.use('brettle:workaround-issue-4970@0.0.1');
  api.use('brettle:accounts-multiple@0.3.1', 'server');
  api.export('AccountsAddService');
  api.addFiles('accounts-add-service-common.js');
  api.addFiles('accounts-add-service-server.js', 'server');
});

Package.onTest(function(api) {
  api.versionsFrom('1.0.4');
  api.use('tinytest');
  api.use('ddp');
  api.use('accounts-base');
  api.use('accounts-password');
  api.use('mongo');
  api.use('brettle:accounts-add-service');
  api.use('brettle:accounts-testing-support@0.4.3');
  api.use('brettle:accounts-multiple@0.3.1', 'server');
  api.addFiles('accounts-add-service-server-tests.js', 'server');
});
